proton/cert_pinning.py

```diff
@@ -0,0 +1,98 @@
+import base64
+import hashlib
+import json
+import os
+
+import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.connectionpool import HTTPSConnectionPool
+from requests.packages.urllib3.poolmanager import PoolManager
+
+crypto = requests.packages.urllib3.contrib.pyopenssl.OpenSSL.crypto
+
+class InspectedHTTPSConnectionPool(HTTPSConnectionPool):
+    """Custom HTTPSConnectionPool that verifies the certificate for each connection"""
+    
+    def _validate_conn(self, conn):
+        r = super(InspectedHTTPSConnectionPool, self)._validate_conn(conn)
+ 
+        sock = conn.sock
+        sock_connection = sock.connection
+
+        if self.is_session_secure(sock_connection.get_peer_cert_chain()[0], conn):
```

Comment:
I would factor out `conn.sock.connection.get_peer_cert_chain()[0]` into a dedicated utility function.
BTW is `[0]` always there?

Reply:
Sounds good! :+1: 
The `[0]` is because `get_peer_cert_chain()` returns a chain of certificates, and [0]. As per docs: [SSL_get_peer_cert_chain() returns a pointer to STACK_OF(X509) certificates forming the certificate chain of the peer.](https://doc.ecoscentric.com/ref/openssl-ssl-ssl-get-peer-cert-chain.html)

Reply:
Is there a chance that the list is empty (and thus you would get an `IndexError`?)

Reply:
Well, I don't think so since the only situation that this could happen would br on a connection timeout (and requests does manage that). But I get your point 👍🏻